python/dgl/nn/pytorch/conv/grouprevres.py

```diff
@@ -0,0 +1,186 @@
+                    x.append(element.detach())
+                    ctx.input_requires_grad.append(element.requires_grad)
+                else:
+                    x.append(element)
+                    ctx.input_requires_grad.append(None)
+            # Detach the output, which then allows discarding the intermediary results
+            outputs = ctx.fn(*x).detach_()
+
+        # clear memory of input node features
+        inputs[0].storage().resize_(0)
+
+        # store for backward pass
+        ctx.inputs = [inputs]
+        ctx.outputs = [outputs]
+
+        return outputs
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        if not torch.autograd._is_checkpoint_valid():
+            raise RuntimeError("InvertibleCheckpoint is not compatible with .grad(), \
+                               please use .backward() if possible")
+        # retrieve input and output tensor nodes
+        if len(ctx.outputs) == 0:
+            raise RuntimeError("Trying to perform backward on the InvertibleCheckpoint \
+                               for more than once.")
+        inputs = ctx.inputs.pop()
+        outputs = ctx.outputs.pop()
+
+        # reconstruct input node features
+        with torch.no_grad():
+            # inputs[0] used to have the input node features
+            inputs_inverted = ctx.fn_inverse(*((outputs,)+inputs[1:]))
+            # clear memory of outputs
+            outputs.storage().resize_(0)
+
+            x = inputs[0]
+            x.storage().resize_(int(np.prod(x.size())))
+            x.set_(inputs_inverted)
+
+        # compute gradients
+        with torch.set_grad_enabled(True):
+            detached_inputs = []
+            for i, element in enumerate(inputs):
+                if isinstance(element, torch.Tensor):
+                    element = element.detach()
+                    element.requires_grad = ctx.input_requires_grad[i]
+                detached_inputs.append(element)
+
+            detached_inputs = tuple(detached_inputs)
+            temp_output = ctx.fn(*detached_inputs)
+
+        filtered_detached_inputs = tuple(filter(lambda x: x.requires_grad, detached_inputs))
+        gradients = torch.autograd.grad(outputs=(temp_output,),
+                                        inputs=filtered_detached_inputs + ctx.weights,
+                                        grad_outputs=grad_outputs)
+
+        input_gradients = []
+        i = 0
+        for rg in ctx.input_requires_grad:
+            if rg:
+                input_gradients.append(gradients[i])
+                i += 1
+            else:
+                input_gradients.append(None)
+
+        gradients = tuple(input_gradients) + gradients[-len(ctx.weights):]
+
+        return (None, None, None) + gradients
+
+
+class GroupRevRes(nn.Module):
+    r"""Grouped reversible residual connections for GNNs, as introduced in
+    `Training Graph Neural Networks with 1000 Layers <https://arxiv.org/abs/2106.07476>`__
+
```

Comment:
Add math description?

Reply:
I will add more description about the general usage of this module. Please refer to torch's doc for group conv2d: https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html
```
Similar to group convolution, `GroupRevRes` applies a group of GNN modules
to the input node features. Given an input value with shape :math:`(V_{src}, D)`, the output
value has shape :math:`(V_{dst}, D)`, where :math:`V_{src}` and :math:`V_{dst}` are the
number of source and destination nodes in the graph, respectively. The feature size
:math:`D` must be divisible by `group`.
```

Reply:
Also, just curious. Does it work with `DGLBlock`?

Reply:
Considering `DGLBlock` was used together with Neighbor Sampling where we only consider <=2 layers, and in RevNet we have thousands of layers. There is no need to support `DGLBlock` IMO.

Reply:
Consider the equation X_i' = f_{w_i}(X_{i-1}', A, U) + X_i. 

For a DGLBlock, the shape of f_{w_i}(X_{i-1}', A, U) is (N_{dst}, D), while X_i is (N_{src}, D), yielding a size mismatch issue. Extending it might be possible, but non-trivial.

Reply:
Updated the description, though different from your proposal.